### PR TITLE
Fix incorrect registration of behavior trees containing faulty XML

### PR DESCRIPTION
**Pull request by schornakj**
This fixes a problem where a behavior tree containing an invalid structure (for example, a BehaviorTree with less than one child or more than two children) could still be instantiated by the factory, even though the XML parser threw and exception when trying to load it. This was possible because adding a BehaviorTree's name to the list of available behavior trees occurred before the check for the validity of the BehaviorTree's XML.

- Modify XML parsing to add a BehaviorTree node to the list of available trees only if that BehaviorTree is valid.
- Delete unused `tree_names` and `tree_count` variables.
- Extend the unit tests to reflect this error case. These tests fail without the fix in this PR.
- ~~Move XML parsing helper functions into an anonymous namespace so they can be used in more places.~~

### CODE REVIEW

File: include/behaviortree_cpp_v3/xml_parsing.h [hunk 1]
```diff
@@ -36,9 +36,6 @@ class XMLParser : public Parser
   Pimpl* _p;
 };
 
-void VerifyXML(const std::string& xml_text,
```
Comment (facontidavide) [inline include/behaviortree_cpp_v3/xml_parsing.h:39]: I think you should not remove this from the public interface

**Comment by facontidavide** [comment]
Thanks for the help.
But I think that you should not remove the function from having public access, since Groot 1.0 relies on that.

Also, seems like you moved some code without much modification, and this makes it very hard for me to understand if you changed something or not.

File: src/xml_parsing.cpp [hunk 2]
```diff
@@ -46,6 +46,188 @@ auto StrEqual = [](const char* str1, const char* str2) -> bool {
...
+    count++;
+  }
+  return count;
+}
+
+/**
+ * @brief Recursively enforces that the attributes of an XML node and all its child nodes are all correct.
+ * @throw BT::RuntimeError if an error was found.
+ */
+void verifyXMLRecursively(
```
Comment (facontidavide) [inline src/xml_parsing.cpp:79]: DO not move this code (hard for me to compare) and keep name and namespace

**Review by facontidavide — CHANGES_REQUESTED** [review]

File: include/behaviortree_cpp_v3/xml_parsing.h (hunk 1 above)
Comment (schornakj) [reply include/behaviortree_cpp_v3/xml_parsing.h]: Understood, I put it back. Good to know that Groot uses this too.

File: src/xml_parsing.cpp (hunk 2 above)
Comment (schornakj) [reply src/xml_parsing.cpp]: I reverted my changes to these functions to keep the inline helper functions, which reduces the size of the diff.

**Pull request merged**